### DISCUSSION TITLE
How can I get the metadata from Terrain3DTextureAsset in c#

### DISCUSSION THREAD

**Discussion by Timit0** [TokisanGames/Terrain3D#557] [Questions] [answered]
# How can I retrieve the metadata of a Terrain3DTextureAsset in C#?
I would like to retrieve the metadata of the Terrain3DTextureAsset after touching Terrain3D with rayCast3D.

![screenShot](https://github.com/user-attachments/assets/6c17cc92-9a5f-4163-a509-16b06cd9fc80)

My current code 

![2](https://github.com/user-attachments/assets/cec8b3d7-5386-4a13-90c0-b7e63ffa9858)

If you can help me, that would be great.

**Comment by TheOrioli** [comment, marked as answer]
Assuming you have access to the Terrain3D node:
```csharp
 var someData = Terrain3D.Get("assets").As<Resource>().Call("get_texture", textureIdx).As<Resource>().Get("whatever you want");
```

**Reply by TokisanGames** [reply]
The full API is available to all languages, including C#, though there are hoops to jump through due to lack of bindings. https://terrain3d.readthedocs.io/en/latest/docs/programming_languages.html

These API pages provide the programmatic path to all variables, as demonstrated by TheOrioli's code.
https://terrain3d.readthedocs.io/en/latest/api/class_terrain3dassets.html

**Reply by TokisanGames** [reply]
For completeness, I think what is missing from all of the answers in regard to the question is to use the raycast and if the terrain is hit, call `get_texture_id(hit_position)`. From there the Terrain3DTextureAsset can be looked up to retrieve whatever meta data is desired. The extended class info below is interesting and valuable. Familiarity with the API and the C# hoops are also necessary.

**Comment by TheOrioli** [comment]
Personally, I've made this extended GDScript class that wraps around the regular TerrainTexture
```gdscript
class_name Terrain3DDataTexture
extends Terrain3DTextureAsset

const TD := preload("res://game/dungeons/dungeon_terrain/terrain_data/DungeonTerrainData.cs")

@export var data : TD
```

which lets me create my own custom data objects that are typesafe in C#
It looks like this in Godot:
![image](https://github.com/user-attachments/assets/ef0ce3f3-53c6-4fa3-a29b-cb4d19e83849)

**Reply by Timit0** [reply]
I triedI tried but I get the error “Line 6:Export type can only be built-in, a resource, a node, or an enum.”
![f](https://github.com/user-attachments/assets/a9221dbb-7483-4343-8a92-cf1cf1ade11e)

**Reply by TheOrioli** [reply]
That depends on what your TerrainData script is. I'm also not sure how Godot handles directory names with special characters like '#', so I would avoid that.

If it looks something like this it should work, and problem is probably specific to your project
```csharp
using Godot;


[GlobalClass]
public partial class DungeonTerrainData : Resource
{
    [Export(PropertyHint.Range, "0,1")] public float Friction { get; private set; } = 0.0f;
}
```

**Reply by Timit0** [reply]
Thx bro 👍👍

**Comment by Timit0** [comment]
To clarify my case

![explanation](https://github.com/user-attachments/assets/c20a9b7a-4f9f-47ef-8220-18743c4f3b13)

Red arrow are RayCast3D
